### PR TITLE
cosma: add new versions and improve package

### PR DESCRIPTION
**Pull request by haampie**
Ping @kabicm

### CODE REVIEW

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 1]
```diff
@@ -7,61 +7,92 @@
...
     version('2.2.0', sha256='1eb92a98110df595070a12193b9221eecf9d103ced8836c960f6c79a2bd553ca')
     version('2.0.7', sha256='8d70bfcbda6239b6a8fbeaca138790bbe58c0c3aa576879480d2632d4936cf7e')
     version('2.0.2', sha256='4f3354828bc718f3eef2f0098c3bdca3499297497a220da32db1acd57920c68d')
 
-    variant('scalapack', default=False,
-            description='Build with ScaLAPACK API.')
+    # We just need the libraries of cuda and rocm, so no need to extend
+    # CudaPackage or ROCmPackage.
+    variant('cuda', default=False, description='Build with cuBLAS support')
+    variant('rocm', default=False, description='Build with rocBLAS support')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cosma/package.py:34]: Can you please elaborate on why there is no need to extend the built-in base packages?

Comment (haampie) [reply var/spack/repos/builtin/packages/cosma/package.py:34]: Hey Tamara, the reason is that cosma just needs the libraries to link to, but not the CUDA or HIP language or compilers for those.

**Comment by tldahlgren** [comment]
@tgamblin @becker33 Any thoughts on this approach?

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 2]
```diff
@@ -7,61 +7,92 @@
...
-    url = 'https://github.com/eth-cscs/COSMA/releases/download/v2.2.0/cosma.tar.gz'
+    url = 'https://github.com/eth-cscs/COSMA/releases/download/v2.5.1/COSMA-v2.5.1.tar.gz'
     git = 'https://github.com/eth-cscs/COSMA.git'
 
+    def url_for_version(self, version):
+        if version <= Version('2.3.0'):
+            return 'https://github.com/eth-cscs/COSMA/releases/download/v{0}/cosma.tar.gz'.format(version)
+
+        return 'https://github.com/eth-cscs/COSMA/releases/download/v{0}/COSMA-v{1}.tar.gz'.format(version, version)
+
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cosma/package.py]: Please move this method below the `depends_on` directives.

File: var/spack/repos/builtin/packages/cosma/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/cosma/package.py:34]: Okay. But inheriting from `CudaPackage` and `ROCmPackage` gives you access to the standard variants, known/supported targets, version- and arch-specific dependencies, and known conflicts.

**Review by tldahlgren — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/cosma/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/cosma/package.py:34]: BUT I have been schooled on why it doesn't matter for you. :wink:

**Comment by tgamblin** [comment]
anything left to do on this one?

**Comment by tldahlgren** [comment]
> anything left to do on this one?

IMO `def url_for_version` should be moved below the `depends_on` directives.  But I have noticed lately that the  placement here has been allowed for a number of packages.

**Comment by haampie** [comment]
Thanks @tldahlgren :D

**Comment by haampie** [comment]
Hi @tldahlgren, can this be merged?

File: var/spack/repos/builtin/packages/cosma/package.py [hunk 3]
```diff
@@ -7,61 +7,92 @@
...
-        args = ['-DCOSMA_WITH_TESTS=OFF',
-                '-DCOSMA_WITH_APPS=OFF',
-                '-DCOSMA_WITH_PROFILING=OFF',
-                '-DCOSMA_WITH_BENCHMARKS=OFF']
 
-        if '^mkl' in spec:
-            args += ['-DCOSMA_BLAS=MKL']
+        if '~scalapack' in spec:
+            return 'OFF'
+        elif '^intel-mkl' in spec or '^intel-oneapi-mkl' in spec:
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cosma/package.py:83]: Interesting.  I didn't realize you could specify a dependency not represented in the package's `depends_on` directives.

**Review by tldahlgren — APPROVED** [review]
Confirmed sha256 for all of the versions.

**Pull request merged**